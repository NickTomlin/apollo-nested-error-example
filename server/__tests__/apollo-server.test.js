"use strict";

const { resolvers, typeDefs } = require("../graphql");
const { ApolloServer } = require("apollo-server");
const { createTestClient } = require("apollo-server-testing");
const gql = require("graphql-tag");

const GET_CUSTOMER_ORDERS = gql`
  query getCustomerOrders($customerId: ID!) {
    customer(id: $customerId) {
      id
      wishlists {
        id
        name
        owner {
          name
        }
      }
      orders {
        amount
        id
        products {
          price
          name
        }
      }
    }
  }
`;

describe("basic service", () => {
  let server;

  beforeEach(async () => {
    server = new ApolloServer({
      typeDefs,
      resolvers
    });
  });

  test("smoke test", async () => {
    const { query } = createTestClient(server);
    const res = await query({
      query: GET_CUSTOMER_ORDERS,
      variables: { customerId: "customer-1234" }
    });
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBeUndefined();
  });

  test("error handling", async () => {
    const { query } = createTestClient(server);
    const res = await query({
      query: GET_CUSTOMER_ORDERS,
      variables: { customerId: "customer-6666" }
    });
    const [errorOne, errorTwo] = res.errors;

    expect(errorOne.path).toEqual(["customer", "orders", 0, "amount"]);
    expect(errorOne.message).toEqual(
      "There was an issue talking to ProductService for product id product-6666"
    );
    expect(errorOne.extensions.exception).toEqual(
      expect.objectContaining({
        service: "product-service",
        owner: "product-team",
        support: "#product-support",
        additionalData: {
          requestId: "xyz-123"
        }
      })
    );
    expect(errorTwo.path).toEqual(["customer", "orders", 0, "products"]);
    expect(errorTwo.extensions.exception).toEqual(
      expect.objectContaining({
        service: "product-service",
        owner: "product-team",
        support: "#product-support",
        additionalData: {
          requestId: "xyz-123"
        }
      })
    );
  });
});
