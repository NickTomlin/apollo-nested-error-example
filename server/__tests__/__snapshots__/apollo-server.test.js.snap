// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic service smoke test 1`] = `
Object {
  "customer": Object {
    "id": "customer-1234",
    "orders": Array [
      Object {
        "amount": 7.98,
        "id": "order-1234",
        "products": Array [
          Object {
            "name": "Chain",
            "price": 2.99,
          },
          Object {
            "name": "Spoke",
            "price": 4.99,
          },
        ],
      },
    ],
    "wishlists": Array [
      Object {
        "id": "wishlist-1234",
        "name": "Birthday items",
        "owner": Object {
          "name": "Jane",
        },
      },
    ],
  },
}
`;
